splice_si_api/src/main/java/com/splicemachine/si/impl/store/IgnoreTxnSupplierImpl.java

```diff
@@ -83,11 +89,41 @@ private void populateIgnoreTxnCache() throws IOException {
                         MultiFieldDecoder decoder = entryDecoder.getEntryDecoder();
                         long startTxnId = decoder.decodeNextLong();
                         long endTxnId = decoder.decodeNextLong();
-                        cache.add(new Pair<Long, Long>(startTxnId, endTxnId));
+                        newCache.add(new Pair<Long, Long>(startTxnId, endTxnId));
                     }
                 }
             }
+            newCache = combineOverlappingRanges(newCache);
+        }
+        return newCache;
+    }
+
+    static Set<Pair<Long, Long>> combineOverlappingRanges(Set<Pair<Long, Long>> cache) {
+        LOG.info("Starting combineOverlappingRanges");
+        BiPredicate<Pair<Long, Long>, Pair<Long, Long>> disjoint = (a, b) -> 
+            a.getFirst() >= b.getSecond() || a.getSecond() <= b.getFirst();
+        BiFunction<Pair<Long, Long>, Pair<Long, Long>, Pair<Long, Long>> combine = (a, b) ->
+            new Pair<Long, Long>(
+                a.getFirst() < b.getFirst() ? a.getFirst() : b.getFirst(),
+                a.getSecond() > b.getSecond() ? a.getSecond() : b.getSecond()
+            );
+        Set<Pair<Long, Long>> newCache = new HashSet<>();
+        Set<Pair<Long, Long>> discard = new HashSet<>();
+        List<Pair<Long, Long>> ranges = new ArrayList<>(cache);
+        while( ! ranges.isEmpty() ) {
+            Pair<Long, Long> r0 = ranges.remove(0);
+            for( Pair<Long, Long> r : ranges ) {
+                if( ! disjoint.test(r, r0) ) {
+                    r0 = combine.apply(r, r0);
+                    discard.add(r);
+                }
+            }
+            newCache.add(r0);
+            ranges.removeAll( discard );
+            discard.clear();
         }
+        LOG.info("Completed combineOverlappingRanges");
+        return newCache;
     }
```

Comment:
(just copied my comment from [here](https://github.com/splicemachine/spliceengine/pull/5103#discussion_r590202496) to make sure it is not lost :) )

I think we can reduce the complexity of the overlapping algorithm by maintaining the cache online, i.e. whenever we add a new element, we change the cache immediately, provided that we have a suitable equality / hash function and we use a TreeMap, we can probably reduce the complexity to from quadratic to logarithmic. If there is a JIRA open for this, please refer me to it so we can follow up there.

Reply:
Thanks @hatyo I've added a Jira: https://splicemachine.atlassian.net/browse/DB-11643